File: tests/test_navigation_ways.py
import json
import unittest
from pathlib import Path

from processing import navigation_ways as nav_ways
from processing.build_map import (
    is_active_source_feature,
    resolve_navigation_guidance,
    source_segments_from_geojson,
)


ROOT = Path(__file__).resolve().parents[1]


class NavigationWaysBuildTest(unittest.TestCase):
    def test_reference_ride_memberships_resolve_without_name_inference(self):
        source = json.loads((ROOT / "data/map-source.geojson").read_text())
        registry = json.loads((ROOT / "data/navigation-ways.json").read_text())
        segments = resolve_navigation_guidance(
            source_segments_from_geojson(source),
            registry,
        )

        self.assertEqual(
            segments["שביל תל חי"]["guidance"]["guidanceIdentity"],
            "way:tel-hai-trail",
        )
        self.assertEqual(
            segments["כביש 9974"]["guidance"]["guidanceIdentity"],
            segments["כביש 9974 כפר יובל"]["guidance"]["guidanceIdentity"],
        )
        self.assertEqual(
            segments["שביל אופניים 99 כפר יובל"]["guidance"]["guidanceIdentity"],
            segments["שביל אופניים 99 מעיין ברוך"]["guidance"]["guidanceIdentity"],
        )
        self.assertNotEqual(
            segments["כביש 99 קריית שמונה"]["guidance"]["guidanceIdentity"],
            segments["שביל אופניים 99 כפר יובל"]["guidance"]["guidanceIdentity"],
        )

    def test_unknown_way_is_rejected(self):
        with self.assertRaisesRegex(ValueError, "unknown way"):
            resolve_navigation_guidance(
                {"Broken": {"id": 1, "guidance": {"role": "named-way", "wayId": "missing"}}},
                {"schemaVersion": 1, "enforcement": "migration", "ways": {}},
            )


if __name__ == "__main__":
    unittest.main()


class NavigationWaysSharedFixtureTest(unittest.TestCase):
    """Parity with packages/core/src/data/navigationWayValidation.js.

    Both implementations run the same fixture file, so editor validation and
    build validation cannot drift.
    """

    @classmethod
    def setUpClass(cls):
        cls.fixtures = json.loads(
            (ROOT / "tests/fixtures/navigation-way-names/schema-cases.json").read_text()
        )

    def _assert_issues(self, actual, expected, label):
        self.assertEqual(
            sorted(entry["code"] for entry in actual),
            sorted(entry["code"] for entry in expected),
            label,
        )
        for want in expected:
            found = next(entry for entry in actual if entry["code"] == want["code"])
            for key, value in want.items():
                if key == "code":
                    continue
                self.assertEqual(found.get(key), value, f"{label}: {want['code']}.{key}")

    def test_registry_cases(self):
        for case in self.fixtures["registryCases"]:
            with self.subTest(case["name"]):
                issues, _ = nav_ways.validate_registry(case["registry"])
                self._assert_issues(issues, case["expectedIssues"], case["name"])

    def test_segment_cases(self):
        for case in self.fixtures["segmentCases"]:
            with self.subTest(case["name"]):
                ways = {way_id: {"wayId": way_id} for way_id in case.get("knownWayIds", [])}
                issues, _, reviewed = nav_ways.validate_segment_guidance(
                    case["guidance"],
                    segment_id=case["segmentId"],
                    internal_name=case["internalName"],
                    ways=ways,
                )
                self._assert_issues(issues, case["expectedIssues"], case["name"])
                self.assertEqual(reviewed, case["expectedReviewed"], case["name"])

    def test_structure_cases(self):
        for case in self.fixtures["structureCases"]:
            with self.subTest(case["name"]):
                adjacency = {
                    int(key): {int(value) for value in values}
                    for key, values in (case.get("adjacency") or {}).items()
                }
                evidence = {
                    int(key): value
                    for key, value in (case.get("memberEvidence") or {}).items()
                }
                result = nav_ways.review_way_structure(
                    way_id=case["wayId"],
                    way_kind=case["wayKind"],
                    member_ids=case["memberIds"],
                    adjacency=adjacency,
                    member_evidence=evidence,
                    acknowledged_issue_fingerprints=case.get(
                        "acknowledgedIssueFingerprints", []
                    ),
                    parallel_pairs=case.get("parallelPairs", []),
                )
                self._assert_issues(result["issues"], case["expectedIssues"], case["name"])
                self.assertEqual(len(result["components"]), case["expectedComponentCount"])
                self.assertEqual(result["maxDegree"], case["expectedMaxDegree"])

    def test_fingerprints_match_javascript_format(self):
        self.assertEqual(
            nav_ways.structure_issue_fingerprint(
                "way-structure-branching", "x", {"branchNodes": [2, 1], "maxDegree": 3}
            ),
            "way-structure-branching|x|branchNodes=1,2|maxDegree=3",
        )

    def test_material_parallel_detector(self):
        road = [[35.6 + i * 0.0003, 33.2] for i in range(21)]
        cycleway = [[35.6 + i * 0.0003, 33.200135] for i in range(21)]
        parallel = nav_ways.detect_material_parallel(road, cycleway)
        self.assertIsNotNone(parallel)
        self.assertGreater(parallel["overlapMeters"], 400)

        crossing = [[35.603, 33.19 + i * 0.001] for i in range(21)]
        self.assertIsNone(nav_ways.detect_material_parallel(road, crossing))

        continuation = [[35.606 + i * 0.0003, 33.2] for i in range(21)]
        self.assertIsNone(nav_ways.detect_material_parallel(road, continuation))


class NavigationWaysReportTest(unittest.TestCase):
    def test_current_network_report_shape(self):
        source = json.loads((ROOT / "data/map-source.geojson").read_text())
        registry = json.loads((ROOT / "data/navigation-ways.json").read_text())
        report = nav_ways.build_navigation_ways_report(
            source, registry, is_active=is_active_source_feature
        )
        self.assertEqual(report["schemaVersion"], 1)
        self.assertGreater(report["activeSegments"], 0)
        self.assertEqual(
            report["activeSegments"],
            report["reviewedSegments"] + report["unreviewedSegments"],
        )
        # Migration mode: unreviewed segments are reported, never blocking.
        self.assertEqual(report["blockingCount"], 0)
        self.assertFalse(report["coverageComplete"])

        summary = nav_ways.manifest_guidance_summary(report)
        self.assertEqual(
            set(summary),
            {
                "schemaVersion",
                "enforcement",
                "activeSegments",
                "reviewedSegments",
                "coverageComplete",
                "conflictCount",
            },
        )

    def test_required_mode_blocks_unclassified_segments(self):
        source = json.loads((ROOT / "data/map-source.geojson").read_text())
        registry = json.loads((ROOT / "data/navigation-ways.json").read_text())
        report = nav_ways.build_navigation_ways_report(
            source,
            {**registry, "enforcement": "required"},
            is_active=is_active_source_feature,
        )
        self.assertGreater(report["blockingCount"], 0)


if __name__ == "__main__":
    unittest.main()

File: tests/test-navigation-ways.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

import {
  deriveGuidanceIndexes,
  fallbackGuidanceKind,
  guidanceClassLabel,
  guidanceModeForSchema,
  isSupportedGuidanceSchema,
  normalizeResolvedSegmentGuidance,
} from "@cycleways/core/data/navigationWays.js";
import {
  detectMaterialParallel,
  reviewWayStructure,
  structureIssueFingerprint,
  validateRegistry,
  validateSegmentGuidance,
} from "@cycleways/core/data/navigationWayValidation.js";

const fixtures = JSON.parse(
  readFileSync(
    fileURLToPath(new URL("./fixtures/navigation-way-names/schema-cases.json", import.meta.url)),
    "utf8",
  ),
);

function summarize(issues) {
  return issues.map((entry) => entry.code).sort();
}

function expectMatch(actual, expected, label) {
  assert.deepEqual(summarize(actual), summarize(expected), `${label}: issue codes`);
  for (const want of expected) {
    const found = actual.find((entry) => entry.code === want.code);
    assert.ok(found, `${label}: missing ${want.code}`);
    for (const [key, value] of Object.entries(want)) {
      if (key === "code") continue;
      assert.equal(found[key], value, `${label}: ${want.code}.${key}`);
    }
  }
}

// --- registry -------------------------------------------------------------
for (const testCase of fixtures.registryCases) {
  const { issues } = validateRegistry(testCase.registry);
  expectMatch(issues, testCase.expectedIssues, `registry/${testCase.name}`);
}

// --- per-segment records --------------------------------------------------
for (const testCase of fixtures.segmentCases) {
  const ways = new Map((testCase.knownWayIds || []).map((wayId) => [wayId, { wayId }]));
  const { issues, reviewed } = validateSegmentGuidance(testCase.guidance, {
    segmentId: testCase.segmentId,
    internalName: testCase.internalName,
    ways,
  });
  expectMatch(issues, testCase.expectedIssues, `segment/${testCase.name}`);
  assert.equal(reviewed, testCase.expectedReviewed, `segment/${testCase.name}: reviewed`);
}

// --- structure review -----------------------------------------------------
for (const testCase of fixtures.structureCases) {
  const adjacency = new Map(
    Object.entries(testCase.adjacency || {}).map(([id, neighbours]) => [
      Number(id),
      new Set(neighbours.map(Number)),
    ]),
  );
  const memberEvidence = new Map(
    Object.entries(testCase.memberEvidence || {}).map(([id, evidence]) => [Number(id), evidence]),
  );
  const result = reviewWayStructure({
    wayId: testCase.wayId,
    wayKind: testCase.wayKind,
    memberIds: testCase.memberIds,
    adjacency,
    memberEvidence,
    acknowledgedIssueFingerprints: testCase.acknowledgedIssueFingerprints || [],
    parallelPairs: testCase.parallelPairs || [],
  });
  expectMatch(result.issues, testCase.expectedIssues, `structure/${testCase.name}`);
  assert.equal(
    result.components.length,
    testCase.expectedComponentCount,
    `structure/${testCase.name}: component count`,
  );
  assert.equal(
    result.maxDegree,
    testCase.expectedMaxDegree,
    `structure/${testCase.name}: max degree`,
  );
}

// A structure warning must never be resolvable by splitting the way: the same
// members under one ID keep one guidance identity, which is the whole point.
{
  const branching = fixtures.structureCases.find((c) => c.expectedMaxDegree === 4);
  const result = reviewWayStructure({
    wayId: branching.wayId,
    wayKind: branching.wayKind,
    memberIds: branching.memberIds,
    adjacency: new Map(
      Object.entries(branching.adjacency).map(([id, n]) => [Number(id), new Set(n.map(Number))]),
    ),
  });
  assert.ok(
    result.issues.every((entry) => entry.severity === "warning"),
    "branching alone never blocks",
  );
}

// --- fingerprints ---------------------------------------------------------
{
  const base = structureIssueFingerprint("way-structure-branching", "x", {
    branchNodes: [1, 2],
    maxDegree: 3,
  });
  assert.equal(
    base,
    structureIssueFingerprint("way-structure-branching", "x", {
      branchNodes: [2, 1],
      maxDegree: 3,
    }),
    "fingerprints are order-independent",
  );
  assert.notEqual(
    base,
    structureIssueFingerprint("way-structure-branching", "x", {
      branchNodes: [1, 2, 5],
      maxDegree: 3,
    }),
    "a membership change yields a new fingerprint",
  );
}

// --- material-parallel detector ------------------------------------------
{
  // Two lines 15 m apart running 600 m in the same direction: the road and its
  // parallel cycleway. This is the case that must fire.
  const road = [];
  const cycleway = [];
  for (let index = 0; index <= 20; index += 1) {
    const lng = 35.6 + index * 0.0003;
    road.push([lng, 33.2]);
    cycleway.push([lng, 33.200135]);
  }
  const parallel = detectMaterialParallel(road, cycleway);
  assert.ok(parallel, "a sustained 15 m corridor is material");
  assert.ok(parallel.overlapMeters > 400);
  assert.ok(parallel.separationMeters < 25);

  // Two lines crossing at right angles must not fire it.
  const crossing = [];
  for (let index = 0; index <= 20; index += 1) {
    crossing.push([35.603, 33.19 + index * 0.001]);
  }
  assert.equal(detectMaterialParallel(road, crossing), null, "a crossing is not parallel");

  // Two chained segments meeting end to end must not fire it either.
  const continuation = [];
  for (let index = 0; index <= 20; index += 1) {
    continuation.push([35.606 + index * 0.0003, 33.2]);
  }
  assert.equal(
    detectMaterialParallel(road, continuation),
    null,
    "an end-to-end continuation is not parallel",
  );
}

// --- presentation fallbacks ----------------------------------------------
{
  assert.equal(guidanceClassLabel("dirt-road"), "דרך עפר");
  assert.equal(guidanceClassLabel(null, "track"), "דרך עפר");
  assert.equal(guidanceClassLabel("connector"), "מקטע מקשר");
  // Nothing reaches the generic phrase because the table lacked an entry: it is
  // the floor for spans with neither a kind nor a route class.
  assert.equal(guidanceClassLabel(null, null), "המשך במסלול");
  assert.equal(fallbackGuidanceKind(null, "secondary"), "road");
  assert.equal(fallbackGuidanceKind(null, "unknown-class"), "other");
  assert.equal(guidanceClassLabel(fallbackGuidanceKind(null, "unknown-class")), "מקטע");
}

// --- schema-bound guidance mode ------------------------------------------
{
  assert.equal(isSupportedGuidanceSchema(1), true);
  assert.equal(isSupportedGuidanceSchema(2), false);
  assert.equal(guidanceModeForSchema(1), "guidance-v1");
  assert.equal(guidanceModeForSchema(null), "legacy");
  assert.equal(guidanceModeForSchema(2), "legacy");
}

// --- runtime indexes ------------------------------------------------------
{
  const segmentsData = {
    "כביש 99 קריית שמונה": {
      id: 174,
      guidance: {
        role: "named-way",
        wayId: "road-99",
        guidanceIdentity: "way:road-99",
        name: "כביש 99",
        kind: "road",
      },
    },
    "כביש 99 שאר ישוב": {
      id: 61,
      guidance: {
        role: "named-way",
        wayId: "road-99",
        guidanceIdentity: "way:road-99",
        name: "כביש 99",
        kind: "road",
      },
    },
    "גשר עינות ירדן": {
      id: 14,
      guidance: {
        role: "standalone",
        guidanceIdentity: "standalone:14",
        name: "גשר עינות ירדן",
        kind: "bridge",
      },
    },
    "שדות דפנה צפון": { id: 244 },
  };
  const { bySegmentId, membersByWayId, issues } = deriveGuidanceIndexes(segmentsData);
  assert.deepEqual(issues, []);
  assert.equal(bySegmentId.size, 3, "unreviewed segments are absent from the index");
  assert.equal(bySegmentId.get(244), undefined);
  assert.deepEqual(membersByWayId.get("road-99").segmentIds, [61, 174]);
  assert.equal(membersByWayId.has("standalone:14"), false, "standalone features are not ways");
}

// A duplicate stable ID would make guidance resolution order-dependent.
{
  const { issues } = deriveGuidanceIndexes({ a: { id: 1 }, b: { id: 1 } });
  assert.deepEqual(summarize(issues), ["segment-duplicate-stable-id"]);
}

// A name or spoken-name edit never changes guidance identity.
{
  const before = normalizeResolvedSegmentGuidance(
    { role: "named-way", wayId: "road-99", guidanceIdentity: "way:road-99", name: "כביש 99", kind: "road" },
    174,
  );
  const after = normalizeResolvedSegmentGuidance(
    {
      role: "named-way",
      wayId: "road-99",
      guidanceIdentity: "way:road-99",
      name: "כביש 99 החדש",
      spokenName: "כביש תשעים ותשע",
      kind: "road",
    },
    174,
  );
  assert.equal(before.guidanceIdentity, after.guidanceIdentity);
}

console.log("test-navigation-ways: OK");
